sycl/doc/design/OffloadDesign.md

```diff
@@ -0,0 +1,326 @@
+files, ELF objects, executables and shared objects, COFF objects, archives or
+simply stored as an offload binary.
+
+We should have the ability to package SPIR-V based device binaries in the
+offload section of any given binary.  These device binaries will be packaged
+as normal with the packager and placed within the given section.
+
+## Clang Linker Wrapper
+
+The clang-linker-wrapper provides the interface to perform the needed link
+steps when consuming fat binaries.  The linker wrapper performs a majority of
+the work involved during the link step during an offload compilation,
+significantly reducing the amount of work that is occuring in the compiler
+driver.  From the compilation perspective, the linker wrapper replaces the
+typical call to the host link.  This allows for the responsibility of the
+compiler driver to be nearly identical when performing a regular compilation
+vs an offloading compilation.
+
+Example usage of the external `clang-linker-wrapper` call:
+
+`clang-linker-wrapper <wrapper opts> -- <linker opts>`
+
+### Device Extraction
+
+During the compilation step, the device binaries are embedded in a section of
+the host binary. When performing the link, this section is extracted from the
+object and mapped according to the device kind.
+
+Depending on the type of binary, the device is embedded as follows:
+
+#### Objects
+
+Object types considered are COFF objects, ELF executables, ELF relocatable
+objects and ELF shared objects.  The device section of the object is in a
+section marked by `.llvm.offloading` for COFF objects.  For ELF files, there is
+a section marked with the `LLVM_OFFLOADING` type.
+
+#### Bitcode
+
+The section which contains the offloading data is from the `.llvm.offloading`
+section within the `llvm.embedded.object` metadata.
+
+#### Offload Binary
+
+The binary itself can be represented by just an offload binary, not requiring
+to be in a section of another binary.  This representation is used for any
+kind of device only binary that is created.
+
+#### Archives
+
+Each item in the archive will be extracted and will be individually checked for
+the file type, properly performing extraction based on the file types already
+listed above.
+
+#### Support of existing fat object format
+
+The current model generates fat objects using the `clang-offload-bundler`
+tool.  This tool creates a binary that contains individual sections that are
+labeled with `__CLANG_OFFLOAD_BUNDLE__kind-triple`.  This format does not
+match the new format as described above.  To enable an ease of transition
+from the old model to the new model, support will be added that allows for
+support of identifying and extracting device binaries that are using the old
+model.
+
+It is expected that all new binaries generated with the updated offloading
+model will represent the embedded fat object format, moving away from the
+`clang-offload-bundler` usage.  We will not support a mixing of fat object
```

Comment:
This could be a problem for users that want to use existing third party libraries with newer libraries. I'm not sure if that's a use case that is worth addressing here.

Reply:
The ability to read in the older format should allow for the mixing of old and new libraries.  It is not expected to be able to have both old and new format objects in a single archive however.

Reply:
That's the situation I'm slightly worried about if a third party vendor doesn't upgrade the format. Is there a mechanism for warning the user when mixing the two types, and is there a way to upgrade an existing library to the newer format?

Reply:
Please note that backwards compatibility is not guaranteed to be preserved forever anyway and there are cases when the application has to be anyway recompiled in order to work with newer runtime.

> It is not expected to be able to have both old and new format objects in a single archive however.

@mdtoguchi, I wonder, though, why can't we support that? If we anyway scan each archive member to handle it in accordance with its type, why can't we detect old/new format at that stage and act on it in a correct way? Am I missing something?

Reply:
We could support, we would need to get down to within the archive level and unbundle/extract on an individual basis there.  Would there be instances where folks would want to use old objects, compile new objects and put them into an archive, or even just adding new objects to an old archive?

Reply:
I think mostly just linking old archives with new archives, but it's a contrived case that we may be able to live without.